File: backend/src/main/java/com/videoplatform/quality/ConnectionQualityService.java
package com.videoplatform.quality;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

/**
 * Classifica a qualidade de conexao a partir de RTT, packet loss e jitter.
 *
 * <p>Regra "pior metrica vence" (Sprint 3 §37): calcula um nivel por metrica
 * disponivel e devolve o pior. Metricas ausentes sao ignoradas.
 *
 * <p>Os limites abaixo sao <b>heuristicas iniciais</b> — nao um padrao universal
 * de qualidade de internet. Devem poder ser ajustados.
 */
@Service
public class ConnectionQualityService {

    private static final Logger log = LoggerFactory.getLogger(ConnectionQualityService.class);

    // limite superior (exclusivo) de cada nivel; acima do ultimo => VERY_POOR
    private static final int[] RTT_MS = {80, 150, 250, 400};
    private static final double[] LOSS_PCT = {1.0, 2.0, 5.0, 10.0};
    private static final int[] JITTER_MS = {15, 30, 50, 100};

    private static final QualityLevel[] BY_INDEX = {
            QualityLevel.EXCELLENT, QualityLevel.GOOD, QualityLevel.FAIR,
            QualityLevel.POOR, QualityLevel.VERY_POOR
    };

    public QualityLevel classify(Integer rttMs, Double packetLossPercent, Integer jitterMs) {
        QualityLevel worst = null;

        if (rttMs != null) {
            worst = worst(worst, levelFor(rttMs, RTT_MS));
        }
        if (packetLossPercent != null) {
            worst = worst(worst, levelFor(packetLossPercent, LOSS_PCT));
        }
        if (jitterMs != null) {
            // jitter e' complementar: sozinho nao rebaixa abaixo de FAIR
            QualityLevel jitterLevel = levelFor(jitterMs, JITTER_MS);
            if (rttMs == null && packetLossPercent == null
                    && jitterLevel.ordinal() < QualityLevel.FAIR.ordinal()) {
                jitterLevel = QualityLevel.FAIR;
            }
            worst = worst(worst, jitterLevel);
        }

        if (worst == null) {
            log.atWarn().addKeyValue("event", "CONNECTION_QUALITY_NO_METRICS")
                    .setMessage("no metrics available to classify, defaulting to FAIR").log();
            return QualityLevel.FAIR;
        }
        return worst;
    }

    private static QualityLevel levelFor(double value, double[] thresholds) {
        for (int i = 0; i < thresholds.length; i++) {
            if (value < thresholds[i]) {
                return BY_INDEX[i];
            }
        }
        return QualityLevel.VERY_POOR;
    }

    private static QualityLevel levelFor(int value, int[] thresholds) {
        for (int i = 0; i < thresholds.length; i++) {
            if (value < thresholds[i]) {
                return BY_INDEX[i];
            }
        }
        return QualityLevel.VERY_POOR;
    }

    private static QualityLevel worst(QualityLevel a, QualityLevel b) {
        if (a == null) {
            return b;
        }
        return a.ordinal() <= b.ordinal() ? a : b;
    }
}

File: frontend/src/services/connectionQualityService.ts
import type { ConnectionMetrics, QualityLevel } from '../types/connectionQuality';

/**
 * Classificacao de qualidade de conexao — modelo da plataforma, sem dependencia
 * do provider de midia. A coleta de metricas especificas do LiveKit e o
 * mapeamento da escala nativa vivem em `video/livekit/livekitMetrics.ts` (a
 * borda). Aqui ficam apenas os limiares e a regra de classificacao.
 *
 * Limites de classificacao (heuristica inicial — nao um padrao universal de
 * qualidade de internet). Espelham `ConnectionQualityService` do backend.
 * Cada valor e' o limite superior (exclusivo) do nivel; acima do ultimo = POOR.
 * Escala da plataforma (Sprint 10 §8), do pior ao melhor:
 * UNKNOWN, POOR, UNSTABLE, GOOD, EXCELLENT.
 */
export const THRESHOLDS = {
  rttMs: [80, 150, 250, 400],
  packetLossPercent: [1, 2, 5, 10],
  jitterMs: [15, 30, 50, 100],
} as const;

const BY_INDEX: QualityLevel[] = ['EXCELLENT', 'GOOD', 'UNSTABLE', 'POOR'];
const ORDER: QualityLevel[] = ['UNKNOWN', 'POOR', 'UNSTABLE', 'GOOD', 'EXCELLENT'];

function levelFor(value: number, thresholds: readonly number[]): QualityLevel {
  for (let i = 0; i < thresholds.length; i++) {
    if (value < thresholds[i]) return BY_INDEX[i];
  }
  return 'POOR';
}

function worst(a: QualityLevel | null, b: QualityLevel): QualityLevel {
  if (a === null) return b;
  return ORDER.indexOf(a) <= ORDER.indexOf(b) ? a : b;
}

/** Classifica pela pior metrica disponivel. Sem metricas -> UNSTABLE (neutro). */
export function classify(m: ConnectionMetrics): QualityLevel {
  let level: QualityLevel | null = null;
  if (m.rttMs != null) level = worst(level, levelFor(m.rttMs, THRESHOLDS.rttMs));
  if (m.packetLossPercent != null) {
    level = worst(level, levelFor(m.packetLossPercent, THRESHOLDS.packetLossPercent));
  }
  if (m.jitterMs != null) {
    let jitter = levelFor(m.jitterMs, THRESHOLDS.jitterMs);
    if (
      m.rttMs == null &&
      m.packetLossPercent == null &&
      ORDER.indexOf(jitter) < ORDER.indexOf('UNSTABLE')
    ) {
      jitter = 'UNSTABLE';
    }
    level = worst(level, jitter);
  }
  return level ?? 'UNSTABLE';
}
